### PR TITLE
Use SMT engine

### PR DESCRIPTION
**Pull request by leonardoalt**
And prepare for smtchecker pragma removal

### CODE REVIEW

**Comment by leonardoalt** [comment]
https://github.com/ethereum/solidity/pull/11202/commits/a33373095235a6e290ee957eb641debb9c4da89a
shows that this also works for upcoming versions

File: test/smtcallback.js [hunk 1]
```diff
@@ -72,10 +72,20 @@ tape('SMTCheckerCallback', function (t) {
       return { error: 'Fake SMT solver error.' };
     };
 
+    var pragmaSMT = '';
+    var settings = {};
+    // `pragma experimental SMTChecker;` was deprecated in 0.8.4
+    if (!semver.gt(solc.semver(), '0.8.3')) {
```
Comment (bshastry) [inline test/smtcallback.js:78]: Perhaps `lte` for better readability instead of not greater than?

File: test/smtcallback.js [hunk 2]
```diff
@@ -152,9 +162,13 @@ tape('SMTCheckerCallback', function (t) {
...
 
-      if (source.includes('// SMTEngine')) {
-        st.skip('Test requires specific SMTChecker engine.');
-        continue;
+      var engine;
+      if (source.includes('// SMTEngine:')) {
+        let re = new RegExp('// SMTEngine: (\\w+)');
+        let m = source.match(re);
+        assert(m !== undefined);
+        assert(m.length >= 2);
```
Comment (bshastry) [inline test/smtcallback.js:170]: If multiple engines are defined, they (second engine onwards) are silently ignored. Should we instead mandate single engine setting only or would that break something?

File: test/smtcallback.js (hunk 1 above)
Comment (leonardoalt) [reply test/smtcallback.js:78]: the problem with that is pre-releases, so `< 0.8.4` actually wouldn't work for the solidity PR itself. That's why here and in other places we do this instead

File: test/smtcallback.js (hunk 2 above)
Comment (leonardoalt) [reply test/smtcallback.js:177]: @bshastry fixed

Comment (leonardoalt) [reply test/smtcallback.js:177]: and added tests for that

File: test/smtcallback.js [hunk 3]
```diff
@@ -152,9 +162,20 @@ tape('SMTCheckerCallback', function (t) {
...
+        let idx = source.indexOf(option);
+        if (source.indexOf(option, idx + 1) !== -1) {
+          st.skip('SMTEngine option given multiple times.');
+          st.end();
+          return;
+        }
+        let re = new RegExp(option + '(\\w+)');
+        let m = source.match(re);
+        assert(m !== undefined);
+        assert(m.length >= 2);
```
Comment (bshastry) [inline test/smtcallback.js:177]: If the early return for multiple SMT engines is not hit, couldn't this assert be stricter i.e., `m.length == 2`?

**Review by bshastry — APPROVED** [review]
Minor comment about possibly stricter assertion. Looks good otherwise, so I'm going to approve.

File: test/smtcallback.js (hunk 3 above)
Comment (leonardoalt) [reply test/smtcallback.js:177]: I don't think so, this array actually has a bunch of info about the match itself, where the first element is the full match, second is the word we were looking for, third is index and so on.

**Comment by leonardoalt** [comment]
@chriseth @axic I think the coverage decrease is due to parts of the tests already being prepared for https://github.com/ethereum/solidity/pull/11202 which is not tested yet by the tests here, but are in that PR. Can one of you merge it? I don't have access

**Pull request merged**